homeassistant/components/hlk_sw16/__init__.py

```diff
@@ -2,139 +2,142 @@
-            )
+    if hass.data.get(DATA_DEVICE_REGISTER, None) is None:
+        hass.data[DATA_DEVICE_REGISTER] = {}
+    if hass.data.get(DATA_DEVICE_LISTENER, None) is None:
+        hass.data[DATA_DEVICE_LISTENER] = {}
+    device = entry.data[CONF_NAME]
+    host = entry.data[CONF_HOST]
+    port = entry.data[CONF_PORT]
 
-            hass.data[DATA_DEVICE_REGISTER][device] = client
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = entry
 
-            # Load platforms
-            hass.async_create_task(
-                async_load_platform(hass, "switch", DOMAIN, (switches, device), config)
-            )
+    @callback
+    def disconnected():
+        """Schedule reconnect after connection has been lost."""
+        _LOGGER.warning("HLK-SW16 %s disconnected", device)
+        async_dispatcher_send(
+            hass, f"hlk_sw16_device_available_{entry.entry_id}", False
+        )
 
-            # handle shutdown of HLK-SW16 asyncio transport
-            hass.bus.async_listen_once(
-                EVENT_HOMEASSISTANT_STOP, lambda x: client.stop()
-            )
+    @callback
+    def reconnected():
+        """Schedule reconnect after connection has been lost."""
+        _LOGGER.warning("HLK-SW16 %s connected", device)
+        async_dispatcher_send(hass, f"hlk_sw16_device_available_{entry.entry_id}", True)
+
+    async def connect():
+        """Set up connection and hook it into HA for reconnect/shutdown."""
+        _LOGGER.info("Initiating HLK-SW16 connection to %s", device)
+
+        client = await create_hlk_sw16_connection(
+            host=host,
+            port=port,
+            disconnect_callback=disconnected,
+            reconnect_callback=reconnected,
+            loop=hass.loop,
+            timeout=CONNECTION_TIMEOUT,
+            reconnect_interval=DEFAULT_RECONNECT_INTERVAL,
+            keep_alive_interval=DEFAULT_KEEP_ALIVE_INTERVAL,
+        )
+
+        hass.data[DATA_DEVICE_REGISTER][entry.entry_id] = client
+
+        # Load entities
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, "switch")
+        )
+
+        # handle shutdown of HLK-SW16 asyncio transport
+        hass.data[DATA_DEVICE_LISTENER][entry.entry_id] = hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, lambda x: client.stop()
+        )
 
-            _LOGGER.info("Connected to HLK-SW16 device: %s", device)
+        _LOGGER.info("Connected to HLK-SW16 device: %s", device)
 
-        hass.loop.create_task(connect())
+    hass.loop.create_task(connect())
 
-    for device in config[DOMAIN]:
-        add_device(device)
     return True
 
 
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    return await hass.config_entries.async_forward_entry_unload(entry, "switch")
+
+
+async def async_remove_entry(hass, entry):
+    """Stop client config entry."""
+    client = hass.data[DATA_DEVICE_REGISTER].pop(entry.entry_id)
```

Comment:
All of this needs to happen on unload instead.  Remote is when we delete the entry.